test/gpu/test_utils.hpp

```diff
@@ -0,0 +1,223 @@
+#ifndef MIGRAPHX_GUARD_TEST_GPU_TEST_UTILS_HPP
+#define MIGRAPHX_GUARD_TEST_GPU_TEST_UTILS_HPP
+
+#include <migraphx/env.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/operators.hpp>
+#include <migraphx/generate.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/cpu/target.hpp>
+#include <migraphx/gpu/target.hpp>
+#include <migraphx/gpu/miopen.hpp>
+#include <migraphx/gpu/hip.hpp>
+#include <migraphx/manage_ptr.hpp>
+#include <migraphx/type_name.hpp>
+#include <migraphx/verify_args.hpp>
+#include <migraphx/instruction.hpp>
+
+#include <miopen/miopen.h>
+
+#include <future>
+#include <thread>
+#include <cmath>
+#include <numeric>
+
+#include <test.hpp>
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wglobal-constructors"
+#endif
+
+MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_TRACE_GPU_COMPILE)
+
+// An improved async, that doesn't block
+template <class Function>
+std::future<typename std::result_of<Function()>::type> detach_async(Function&& f,
+                                                                    bool parallel = true)
+{
+    if(parallel)
+    {
+        using result_type = typename std::result_of<Function()>::type;
+        std::packaged_task<result_type()> task(std::forward<Function>(f));
+        auto fut = task.get_future();
+        std::thread(std::move(task)).detach();
+        return std::move(fut);
+    }
+    return std::async(std::launch::deferred, std::forward<Function>(f));
+}
+
+struct auto_print
+{
+    static void set_terminate_handler(const std::string& name)
+    {
+        static std::string pname;
+        pname = name;
+        std::set_terminate(+[] {
+            std::cout << "FAILED: " << pname << std::endl;
+            try
+            {
+                std::rethrow_exception(std::current_exception());
+            }
+            catch(const std::exception& e)
+            {
+                std::cout << "    what(): " << e.what() << std::endl;
+            }
+            std::cout << std::endl;
+            for(auto&& handle : auto_print::handlers)
+                handle();
+        });
+    }
+    static std::array<std::function<void()>, 2> handlers;
```

Comment:
To get rid of the clang-tidy issues, I think we need to make this a static local variable, and then only access it through a function:

```cpp
static std::array<std::function<void()>, 2>& handlers
{
    static std::array<std::function<void()>, 2> r = {};
    return r;
}
```

Reply:
fixed